tools/encryption/pkg/encryption-helper.go

```diff
@@ -0,0 +1,263 @@
+		if n.Type == html.ElementNode && n.DataAtom == atom.Section {
+			var contentSubSection bool = false
+			var encrypted bool = false
+			for _, a := range n.Attr {
+				if a.Key == "subscriptions-section" && a.Val == "content" {
+					contentSubSection = true
+				} else if a.Key == "encrypted" {
+					encrypted = true
+				}
+			}
+			if contentSubSection && encrypted {
+				encryptedSections = append(encryptedSections, n)
+			}
+		}
+		for cn := n.FirstChild; cn != nil; cn = cn.NextSibling {
+			stack = append(stack, cn)
+		}
+	}
+	return encryptedSections
+}
+
+// Encrypts the content inside of the input "encryptedSections" nodes.
+func encryptAllSections(parsedHtml *html.Node, encryptedSections []*html.Node, kh *keyset.Handle) error {
+	cipher, err := aead.New(kh)
+	if err != nil {
+		return err
+	}
+	for _, node := range encryptedSections {
+		var content []string
+		for c := node.FirstChild; c != nil; c = c.NextSibling {
+			content = append(content, renderNode(c, true))
+			node.RemoveChild(c)
+		}
+		encContent, err := cipher.Encrypt([]byte(strings.Join(content, "")), nil)
+		if err != nil {
+			return err
+		}
+		textNode := &html.Node{Type: html.TextNode, Data: base64.StdEncoding.EncodeToString(encContent)}
+		attrs := []html.Attribute{
+			html.Attribute{Key: "type", Val: "application/octet-stream"},
+			html.Attribute{Key: "ciphertext", Val: ""},
+		}
+		scriptNode := &html.Node{
+			Type:     html.ElementNode,
+			Data:     "script",
+			DataAtom: atom.Script,
+			Attr:     attrs,
+		}
+		node.AppendChild(scriptNode)
+		scriptNode.AppendChild(textNode)
+	}
+	return nil
+}
+
+// Encrypts the document's symmetric key using the input Keyset.
+func encryptDocumentKey(docKeyset, accessRequirement string, ks tinkpb.Keyset) (string, error) {
+	handle, err := keyset.NewHandleWithNoSecrets(&ks)
+	if err != nil {
+		return "", err
+	}
+	he, err := hybrid.NewHybridEncrypt(handle)
+	if err != nil {
+		return "", err
+	}
+	jsonStr := fmt.Sprintf("{\"accessRequirements\": [\"%s\"], \"key\": \"%s\"}", accessRequirement, docKeyset)
+	enc, err := he.Encrypt([]byte(jsonStr), nil)
+	if err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(enc), nil
+}
+
+// Adds the encrypted document key to the output document's head.
+func addEncryptedDocumentKeyToHead(encryptedKey string, parsedHtml *html.Node) error {
+	for n := parsedHtml.FirstChild; n != nil; n = n.NextSibling {
+		if (n.DataAtom == atom.Html) && (len(n.Attr) != 0) {
+			for cn := n.FirstChild; cn != nil; cn = cn.NextSibling {
+				if cn.DataAtom == atom.Head {
+					attrs := []html.Attribute{
+						html.Attribute{Key: "type", Val: "application/json"},
+						html.Attribute{Key: "cryptokeys", Val: ""},
+					}
+					cryptoKeys := &html.Node{
+						Type:     html.ElementNode,
+						Data:     "script",
+						DataAtom: atom.Script,
+						Attr:     attrs,
+					}
+					jsonData := fmt.Sprintf(`{"google.com":"%s"}`, encryptedKey)
```

Comment:
Possibly coming in the followup PRs? But since we are doing one-pass thing here, the document will not be useful w/o publisher-own encrypted key here. And it won't be possible to add it on a second pass - they should all be prepared at the same time.

Reply:
I am looking to do this in a followup PR, yes. I would like to talk to some other folks about how publishers will be providing their public keys. I am thinking it would either be hosted on an SCS like our public key, or  by direct command line insertion.

Reply:
I don't think we will have their keys on our servers. I think, generically, it'd be a URL. But possibly more practically a keystore on the disk?

Reply:
I definitely think we can start with URLs at this time, with `file:` URLs being a special case for now if we need for manual tests. Then we can expand. It's fine if we do it in followup PRs - this PR is already pretty big.

Reply:
I added support for publishers to input their domain and URL to the hosted public key as command line flags. PTAL.